File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* t1 = l1 ;
        ListNode* t2 = l2 ;
        ListNode* dummy = new ListNode(-1);
        ListNode* temp = dummy ;
        int carry = 0 ;
        while(t1 || t2){
            int sum = carry ;
            if(t1)  sum += t1 -> val ;
            if(t2)  sum += t2 -> val ;

            temp -> next = new ListNode(sum % 10) ;
            carry = sum / 10 ;
            temp = temp -> next ;
            
            if(t1)  t1 = t1 -> next ;
            if(t2)  t2 = t2 -> next ;
        }
        dummy = dummy -> next ;
        if(carry != 0){
            temp -> next = new ListNode(carry) ;
        }
        return dummy ;
    }
};

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null || l2 == null)    return null ;
        int carry = 0 ;
        ListNode temp1 = l1 ;
        ListNode temp2 = l2 ;
        ListNode dummy = new ListNode(-1) ;
        ListNode temp = dummy ;
        while(temp1 != null && temp2 != null){
            int sum = temp1.val + temp2.val + carry;
            ListNode newNode = new ListNode(sum%10);;
            temp.next = newNode ;
            temp = temp.next ;
            carry = sum/10 ;
            temp1 = temp1.next ;
            temp2 = temp2.next ;
        }
        while(temp1 != null){
            int sum = temp1.val + carry ;
            ListNode newNode = new ListNode(sum%10);;
            temp.next = newNode ;
            temp = temp.next ;
            carry = sum/10 ;
            temp1 = temp1.next ;
        }
        while(temp2 != null){
            int sum = temp2.val + carry ;
            ListNode newNode = new ListNode(sum%10);;
            temp.next = newNode ;
            temp = temp.next ;
            carry = sum/10 ;
            temp2 = temp2.next ;
        }
        if(carry != 0){
            ListNode newNode = new ListNode(carry) ;
            temp.next = newNode ;
        }
        return dummy.next ;
    }
}